### PR TITLE
Bug 1667051 - Remove `try` perf data older than 6 weeks

### PR DESCRIPTION
**Pull request by ionutgoldan**
**Bug link:** https://bugzilla.mozilla.org/show_bug.cgi?id=1667051

### CODE REVIEW

File: treeherder/model/management/commands/cycle_data.py [hunk 1]
```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
...
+
+                if not PerformanceDatum.objects.filter(
+                    repository_id=signature.repository_id,  # leverages (repository, signature) compound index
+                    signature_id=signature.id,
+                ).exists():
+                    signature.delete()
+        except MaxRuntimeExceeded as ex:
+            logger.warning(ex)
+
+    def _maybe_quit(self):
```
Comment (alexandru-io) [inline treeherder/model/management/commands/cycle_data.py:132]: I would be more specific here, like: `_quite_runtime_exceeded`. `maybe` is pretty vague and gives room for interpretation/assumption.

File: treeherder/model/management/commands/cycle_data.py [hunk 2]
```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
...
+        self._chunk_size = chunk_size
+        self._max_timestamp = datetime.now() - self._cycle_interval
+        self._manager = PerformanceDatum.objects
+
+    def remove(self, using: CursorWrapper):
+        chunk_size = self._find_ideal_chunk_size()
+        using.execute(
+            '''
+            DELETE FROM `performance_datum`
+            WHERE push_timestamp < %s
```
Comment (alexandru-io) [inline treeherder/model/management/commands/cycle_data.py:209]: Giving the name `_max_timestamp` I would use here like:
`WHERE push_timestamp <= %s`

File: treeherder/model/management/commands/cycle_data.py [hunk 3]
```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
...
+            LIMIT %s
+        ''',
+            [self._max_timestamp, chunk_size],
+        )
+
+    def _find_ideal_chunk_size(self) -> int:
+        max_id = self._manager.filter(push_timestamp__gt=self._max_timestamp).order_by('-id')[0].id
```
Comment (alexandru-io) [inline treeherder/model/management/commands/cycle_data.py:216]: Giving at line `219` you used `order_by('-id')` I would say `'-id'` is a typo.

File: treeherder/model/management/commands/cycle_data.py [hunk 4]
```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
...
+        ''',
+            [self.try_repo, self._max_timestamp, chunk_size],
         )
 
+    def _find_ideal_chunk_size(self) -> int:
+        max_id = (
+            self._manager.filter(
+                push_timestamp__gt=self._max_timestamp, repository_id=self.try_repo
+            )
+            .order_by('-id')[0]
```
Comment (alexandru-io) [inline treeherder/model/management/commands/cycle_data.py:264]: same here: typo `'-id'`

**Review by alexandru-io — CHANGES_REQUESTED** [review]
There are some suggestions and typos I might have found. Otherwise, the logic seems fine.

File: treeherder/model/management/commands/cycle_data.py (hunk 3 above)
Comment (ionutgoldan) [reply treeherder/model/management/commands/cycle_data.py:216]: This is not a typo, it's intentional. So I can get the biggest id that still respects the filters.

File: treeherder/model/management/commands/cycle_data.py (hunk 4 above)
Comment (ionutgoldan) [reply treeherder/model/management/commands/cycle_data.py:264]: Again, this is not a typo

File: treeherder/model/management/commands/cycle_data.py [hunk 5]
```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
...
+        max_runtime: timedelta = None,
+        strategies: List[RemovalStrategy] = None,
+        **kwargs,
+    ):
+        super().__init__(chunk_size, sleep_time, is_debug)
+        self.started_at = None
+        self.max_runtime = max_runtime or PerfherderCycler.DEFAULT_MAX_RUNTIME
+        self.strategies = strategies or RemovalStrategy.fabricate_all_strategies(chunk_size)
 
     def cycle(self):
```
Comment (alexandru-io) [inline treeherder/model/management/commands/cycle_data.py:103]: This function describes **how** we are deleting the data, which might not be a straightforward naming. I would describe **what** it deletes, like:
`def delete_older_than(self, timedelta=DEFAULT_MAX_RUNTIME)`
and let the user decide whether he wants to use the default or some of it's own.
But this is just a recommendation.

Comment (ionutgoldan) [reply treeherder/model/management/commands/cycle_data.py:103]: I disagree. The `PerfherderCycler` doesn't really know what it deletes. That's the responsibility of the strategies it encompasses.
And the strategies don't really stick to the idea of deleting something that's older than X time. The strategies employ various garbage collecting techniques which can be pretty different.

**Review by alexandru-io — APPROVED** [review]

**Review by camd — APPROVED** [review]

**Comment by ionutgoldan** [comment]
I'm afraid I have some bad news. I've just tested this on `treeherder-prototype`. It seems like the new strategy that's targeting `try` data is consistently experiencing intermittents. Most likely these cause timeouts.

I *think* it's because the new delete query isn't leveraging the appropriate index. Which means I need to tweak the algorithm a bit

Good news: all of the previous expiring mechanism works without problems. Currently, the worst case sounds like this: (1) every day, we delete all the data we used to delete so far + (2) a couple tens of thousands (instead of tens of millions) + (3) some extra garbage collection, then quit.

Steps (1) & (3) are OK. Step (2) needs to be fixed.

**Comment by ionutgoldan** [comment]
@camd could you review the changes after your last review? I've put them in a single (and the final) commit.

To summarize, the `performance_datum` table has a compound index over the `repository`, `signature` & `push_timestamp` columns (in this order). Previously, I was deleting the data based on repository & push_timestamp. Unfortunately, this wasn't fully leveraging the index.

Thus, I'm now deleting it based on signature also.
For that, I fetch all the signatures from try (which are around 400,000) & go through 10 of them at a time.
For each bulk of 10 signatures, I can now quickly delete the rows I want. When I finish with a bulk, I go to the next one, until I treat all the try signatures.

**Review by camd — APPROVED** [review]
This looks good to me

**Pull request merged**